### PR TITLE
feat(feeshare): Developer Incentives

### PR DESCRIPTION
**Pull request by Reecepbcups**
This PR is in replace of #75  & builds on main properly

### CODE REVIEW

File: custom/auth/ante/ante.go
Comment (ZaradarBH) [inline custom/auth/ante/ante.go:73]: Your providing the wrong BankKeeper type to your Decorator: "cannot use options.BankKeeper (variable of type BankKeeper) as "github.com/classic-terra/core/x/feeshare/ante".BankKeeper value in argument to feehshareante.NewFeeSharePayoutDecorator: BankKeeper does not implement "github.com/classic-terra/core/x/feeshare/ante".BankKeeper (missing method SendCoinsFromModuleToAccount)"

File: custom/auth/ante/ante.go
Comment (ZaradarBH) [inline custom/auth/ante/ante.go:26]: This should be: FeeShareKeeper     feehshareante.FeeShareKeeper

File: x/feeshare/ante/expected_keepers.go
Comment (ZaradarBH) [inline x/feeshare/ante/expected_keepers.go:13]: Add this to your BankKeeper interface to fix the antehandler type issue:

SendCoinsFromModuleToModule(ctx sdk.Context, senderModule string, recipientModule string, amt sdk.Coins) error

This will also make the tests work

Comment (Reecepbcups) [reply x/feeshare/ante/expected_keepers.go:14]: I had modified the custom/auth/expected_keepers.go accidently it seems here. But still believe you are getting a local cache result, because `TestAnteTestSuite/TestIntegrationTaxExemption suite.go:77: test panicked: UnmarshalJSON cannot decode empty bytes` fails for me. But does not fail when I comment out the feeshare decorator.

I've got to look into it more, but have a busy day today. I suspect it's something small or a way the module needs to change to satisfy v44 in some aspect

Comment (ZaradarBH) [reply x/feeshare/ante/expected_keepers.go:14]: I spent most of the night testing and it seems to be related to your custom params usage. As such if you disable the GetParams call in your decorator the tests work. However overriding the stringer method for Params like the other modules does via "option (gogoproto.goproto_stringer) = false;" doesnt seem to change much. Maybe we can jump on a call and debug it together if you have time. you can reach me on discord @ Zaradar#1078

Comment (ZaradarBH) [reply x/feeshare/ante/expected_keepers.go:14]: Also another thing I noticed in the integration test is that you did not account for the extra BankKeeper your adding in @ https://github.com/Reecepbcups/lunc-core/blob/c7eb092c55947d8d30289968790e56f3bc946029/custom/auth/ante/integration_test.go#L143 , however I do believe you can avoid that argument all together if you just align the expected_keeper in custom/auth/ante and x/feeshare/ante :)

File: custom/auth/ante/ante.go
Comment (ZaradarBH) [inline custom/auth/ante/ante.go:27]: I am not super keen on having two different interfaces for the same keeper in the antehandler. Would it not be cleaner to simply add the methods you need the existing interface in custom/auth/ante/expected_keeper.go and have one less HandlerOptions to keep track of? :)

File: proto/terra/feeshare/v1beta1/genesis.proto
Comment (ZaradarBH) [inline proto/terra/feeshare/v1beta1/genesis.proto:19]: Not sure if its a big deal that your not manually implementing the proto stringer method in your Keeper Params?

e.g: https://github.com/classic-terra/core/blob/649f7c4348bc437c6b26bbd9114febef84c7dd3a/x/market/types/params.go#L48

File: proto/terra/feeshare/v1beta1/genesis.proto
Comment (ZaradarBH) [inline proto/terra/feeshare/v1beta1/genesis.proto:18]: Other x/modules seem to put this message contract in the top level  proto file. E.g. https://github.com/classic-terra/core/blob/649f7c4348bc437c6b26bbd9114febef84c7dd3a/proto/terra/market/v1beta1/market.proto#L9 && https://github.com/classic-terra/core/blob/649f7c4348bc437c6b26bbd9114febef84c7dd3a/proto/terra/oracle/v1beta1/oracle.proto#L9

Maybe we should mimic that convention? :)

**Comment by nghuyenthevinh2000** [comment]
@Reecepbcups @ZaradarBH can you merge this into your branch, it fixes simulation nondeterminism

https://github.com/classic-terra/core/tree/feeshare

**Comment by Reecepbcups** [comment]
> Is there a specification for this feature? I cannot review without it.

x/feeshare/spec

File: custom/auth/ante/ante.go
Comment (Reecepbcups) [reply custom/auth/ante/ante.go:27]: I agree, will do. I figured from your other comment you wanted to keep this the same as before. Will fix later today

**Comment by fragwuerdig** [comment]
@inon-man @Reecepbcups @nghuyenthevinh2000 @ZaradarBH 

I looked into the liveness test issue for this PR.

The local testnet nodes don't come up, because the gentxs cannot be delivered on InitChainer (non-ok exit code but panic message is redacted - so no idea what's going on). I have no clue, why that is.

The liveness test script is tricked into believing all testnet containers started properly, because the nodes need 2-3s seconds before they die. So the docker ps command shows them as being started and the script continues. But then the nodes die.

First I thought there is an issue with my PR #121. But when I switch to main (where #121 is present as well), then the testnet nodes come up properly...

Thoughts?

**Comment by nghuyenthevinh2000** [comment]
Thanks for your find ❤️

File: proto/terra/feeshare/v1beta1/genesis.proto
Comment (ZaradarBH) [reply proto/terra/feeshare/v1beta1/genesis.proto:19]: Tests worked. Not a big deal, probably a better pattern long term :)

**Review by ZaradarBH — APPROVED** [review]
Great job m8 and gz  on the lead dev role @ Juno :)

File: x/feeshare/ante/ante.go
Comment (nghuyenthevinh2000) [inline x/feeshare/ante/ante.go:57]: can you create an unit test for FeeSharePayout? it seems to be an important piece that extract information from MsgExecuteContract but is missing unit test.

File: x/feeshare/client/cli/tx.go
Comment (nghuyenthevinh2000) [inline x/feeshare/client/cli/tx.go:111]: seem to be missing [withdrawer] in Use

Comment (ZaradarBH) [reply x/feeshare/client/cli/tx.go]: Nice find. :)

File: x/feeshare/ante/ante.go
Comment (ZaradarBH) [reply x/feeshare/ante/ante.go:57]: Implemented simple test of FeeSharePayoutDecorator

**Comment by fragwuerdig** [comment]
Do NOT merge this branch @ZaradarBH, @nghuyenthevinh2000, @Reecepbcups: Something is going on. I tried different environments to run this branch - failed. There is some runtime issue going on with this feature. The nodes wont start on genesis transactions.

**Comment by nghuyenthevinh2000** [comment]
> Do NOT merge this branch @ZaradarBH, @nghuyenthevinh2000, @Reecepbcups: Something is going on. I tried different environments to run this branch - failed. There is some runtime issue going on with this feature. The nodes wont start on genesis transactions.

thanks you so much for testing this out, what is the log?

**Comment by fragwuerdig** [comment]
@Reecepbcups @ZaradarBH @nghuyenthevinh2000 @inon-man: I request a code change for this branch. Unfortunately I cannot request code changes the "github-way" because I am not in the list of reviewers. In `app.go` beginning at line `522` we have the following snippet:

`app.mm.SetOrderInitGenesis(
		capabilitytypes.ModuleName, authtypes.ModuleName,
		banktypes.ModuleName, distrtypes.ModuleName,
		stakingtypes.ModuleName, slashingtypes.ModuleName,
		govtypes.ModuleName, markettypes.ModuleName,
		oracletypes.ModuleName, treasurytypes.ModuleName,
		wasmtypes.ModuleName, authz.ModuleName,
		minttypes.ModuleName, crisistypes.ModuleName,
		ibchost.ModuleName, genutiltypes.ModuleName,
		evidencetypes.ModuleName, ibctransfertypes.ModuleName,
		feegrant.ModuleName,
		feesharetypes.ModuleName
)`

Please paste the following snippet:

`app.mm.SetOrderInitGenesis(
feesharetypes.ModuleName,
capabilitytypes.ModuleName, authtypes.ModuleName,
		banktypes.ModuleName, distrtypes.ModuleName,
		stakingtypes.ModuleName, slashingtypes.ModuleName,
		govtypes.ModuleName, markettypes.ModuleName,
		oracletypes.ModuleName, treasurytypes.ModuleName,
		wasmtypes.ModuleName, authz.ModuleName,
		minttypes.ModuleName, crisistypes.ModuleName,
		ibchost.ModuleName, genutiltypes.ModuleName,
		evidencetypes.ModuleName, ibctransfertypes.ModuleName,
		feegrant.ModuleName
)`

This moves the `feesharetypes.ModuleName` up in the `SetOrderInitGenesis`. After setting higher priority for genesis initialization of the feeshare module the node should start normally on genesis (tested it locally, produced some blocks).

It might very well be that the liveness test fails nevertheless. But then its not because the node fails to start up but rather because expanding the local testnet from 4 to 6 nodes introduced a bug with this test. I guess, we should work on it in another branch.

Btw: Do we need migration handler for this? On the mainnet the parameter space for this is not yet initialized, or is it?

**Comment by fragwuerdig** [comment]
> > Do NOT merge this branch @ZaradarBH, @nghuyenthevinh2000, @Reecepbcups: Something is going on. I tried different environments to run this branch - failed. There is some runtime issue going on with this feature. The nodes wont start on genesis transactions.
> 
> thanks you so much for testing this out, what is the log?

The nodes simply did not start up on genesis:

grafik

**Comment by ZaradarBH** [comment]
> Do NOT merge this branch @ZaradarBH, @nghuyenthevinh2000, @Reecepbcups: Something is going on. I tried different environments to run this branch - failed. There is some runtime issue going on with this feature. The nodes wont start on genesis transactions.

> @Reecepbcups @ZaradarBH @nghuyenthevinh2000 @inon-man: In `app.go` beginning at line `522` we have the following snippet:
> 
> `app.mm.SetOrderInitGenesis( capabilitytypes.ModuleName, authtypes.ModuleName, banktypes.ModuleName, distrtypes.ModuleName, stakingtypes.ModuleName, slashingtypes.ModuleName, govtypes.ModuleName, markettypes.ModuleName, oracletypes.ModuleName, treasurytypes.ModuleName, wasmtypes.ModuleName, authz.ModuleName, minttypes.ModuleName, crisistypes.ModuleName, ibchost.ModuleName, genutiltypes.ModuleName, evidencetypes.ModuleName, ibctransfertypes.ModuleName, feegrant.ModuleName, feesharetypes.ModuleName )`
> 
> This must be replaced by the following snippet:
> 
> `app.mm.SetOrderInitGenesis( feesharetypes.ModuleName, capabilitytypes.ModuleName, authtypes.ModuleName, banktypes.ModuleName, distrtypes.ModuleName, stakingtypes.ModuleName, slashingtypes.ModuleName, govtypes.ModuleName, markettypes.ModuleName, oracletypes.ModuleName, treasurytypes.ModuleName, wasmtypes.ModuleName, authz.ModuleName, minttypes.ModuleName, crisistypes.ModuleName, ibchost.ModuleName, genutiltypes.ModuleName, evidencetypes.ModuleName, ibctransfertypes.ModuleName, feegrant.ModuleName )`
> 
> This moves the `feesharetypes.ModuleName` up in the `SetOrderInitGenesis`. After setting higher priority for genesis initialization of the feeshare module the node should start normally on genesis (tested it locally, produced some blocks).
> 
> It might very well be that the liveness test fails nevertheless. But then its not because the node fails to start up but rather because expanding the local testnet from 4 to 6 nodes introduced a bug with this test. I guess, we should work on it in another branch.
> 
> Btw: Do we need migration handler for this? On the mainnet the parameter space for this is not yet initialized, or is it?
> 
> Another question: Why is only one workflow (snyk test) running on this branch?

Nice find m8! I think the workflow issue might be because the PR branch belongs to an external org, it might be messing with the GH action configuration. Havnt looked into it.

FYI I created an issue for the liveliness bug so we can all look into that together next week. #171

**Comment by fragwuerdig** [comment]
@Reecepbcups I have another commit for this branch (merging main into this branch to resolve conflicts). I don't know exactly why I could push https://github.com/classic-terra/core/pull/144/commits/ed35c8c67fffa660136d7ca4d2e451c6f703955f. But it's clear that I cannot push anymore... Do you changed any permissions on your forked repo? Can you allow me to push commits?

**Comment by Reecepbcups** [comment]
@fragwuerdig Thanks for this! I missed running a local instance I got so caught up in my Juno upgrade.

I did not change any permissions no. I have allow edits by maintainers on and have not touched anything since this PR was opened. So unless this org changed permissions, not sure why this is the case

**Review by nghuyenthevinh2000 — APPROVED** [review]
Code looks good and best to merge this after sdk 45 upgrade

**Pull request merged**
